p8e-util/src/main/kotlin/io/p8e/crypto/Pen.kt

```diff
@@ -64,19 +64,61 @@ class Pen(
             .orThrow { IllegalStateException("can't verify signature - public cert may not match private key.") }
     }
 
-    override fun sign(): ByteArray = signature.sign()
-
-    override fun update(data: ByteArray) { signature.update(data) }
+    override fun sign(): ByteArray {
+        signature.update(aggregatedData)
+        return signature.sign()
+    }
 
-    override fun update(data: ByteArray, off: Int, len: Int) { signature.update(data, off, len) }
+    override fun update(data: ByteArray) = signature.update(data)
+
+    override fun update(data: ByteArray, off: Int, res: Int) {
+        // If off is less then res, these are the data that we care about.
+        if(off < res) {
+            if(!verifying) {
+                val dataSample = data.copyOfRange(off, off+res)
+                aggregatedData = dataSample
+            } else {
+                /**
+                 * The downstream (data verification) chunks the data into data size of 8192.
+                 * The data needs to be aggregated to its signing size of 32768 before the
+                 * data can be validated.
+                 */
+                if(objSizeIndexer == OBJECT_SIZE_BYTES) {
+                    objSizeIndexer = (off + res)
+                    aggregatedData = null
+                    aggregatedData = if (aggregatedData == null) {
+                        data.copyOfRange(off, off + res)
+                    } else {
+                        aggregatedData?.plus(data.copyOfRange(off, off + res))
+                    }
+                } else {
+                    objSizeIndexer += (off + res)
+                    aggregatedData = aggregatedData?.plus(data.copyOfRange(off, off + res))
+                }
+            }
+        }
+    }
 
     override fun update(data: Byte) { signature.update(data) }
 
-    override fun verify(signatureBytes: ByteArray): Boolean = signature.verify(signatureBytes)
+    override fun verify(signatureBytes: ByteArray): Boolean {
+        signature.update(aggregatedData)
 
-    override fun initVerify(publicKey: PublicKey) { signature.apply { initVerify(publicKey) } }
+        // Reset the object size indexer.
+        objSizeIndexer = OBJECT_SIZE_BYTES
 
-    override fun initSign() { signature.apply { initSign(keyPair.private) } }
+        return signature.verify(signatureBytes)
+    }
+
+    override fun initVerify(publicKey: PublicKey) {
+        signature.apply { initVerify(publicKey) }
```

Comment:
I think this could just be `signature.initVerify(publicKey)`, right? The apply seems unnecessary...

Reply:
Yea, I'll clean this up